src/libcollections/vec.rs

```diff
@@ -1322,8 +1322,10 @@ impl<T> Drop for Vec<T> {
         // OK because exactly when this stops being a valid assumption, we
         // don't need unsafe_no_drop_flag shenanigans anymore.
         if self.buf.unsafe_no_drop_flag_needs_drop() {
-            for x in self.iter_mut() {
-                unsafe { drop_in_place(x); }
+            if unsafe { needs_drop::<T>() } {
+                for x in self.iter_mut() {
+                    unsafe { drop_in_place(x); }
+                }
```

Comment:
Can you change the `unsafe` block to wrwap the whole `if needs_drop` block here? This is the preferred style.